app/src/test/java/org/mozilla/tv/firefox/fxa/FxaReceivedTabTest.kt

```diff
@@ -25,9 +36,10 @@ class FxaReceivedTabTest {
         val inputTabData = expectedTabUrls.mapIndexed { i, url -> TabData("tab title $i", url) }
         val tabReceivedEvent = mockADMTabReceivedEvent(DeviceType.DESKTOP, expectedDeviceName, inputTabData)
 
-        receivedTabsRaw.onNext(tabReceivedEvent)
-
-        receivedTabsTestObs.assertValues(expected)
+        Observable.just(tabReceivedEvent)
```

Comment:
IMO, the valuable thing here is verifying that the transformation is happening as intended.  The rest of it is really just "verify that this method was called", which is fine, but feels like a bit of a waste of time to me.  Both to write, and also to slow down future changes.  I don't think it catches real mistakes.  If you strongly disagree, we can follow up and add some additional tests in.

I did, however, update the names.

Reply:
From a theoretical perspective, my guiding idea is that tests should prevent people from accidentally changing functionality. For example, if someone is reordering lines, I don't want them to forget to add a line back in. There's an argument that this should be caught in review but these types of bugs are often missed in reviews. Therefore, if devs are actually changing behavior, I expect them to have to change the tests too.

>  also to slow down future changes.

I've never been on a team where there were so many tests that they got in the way of making changes so this argument doesn't concern me yet.

---

For our particular case, we want to make sure "send tabs" is tested. The side taking raw push events and converting them into domain objects is trivial (so maybe not worth testing?) and the view layer is hard to test. The original test was intended to do the primary chunk of testing to ensure we're actually exposing the events we intend to outside of the model. However, there are no automated tests for that now. Therefore, I'd rather have a single test that tests both the integration, and transitively the transformation function, than one test that tests the transformation which may accidentally stop being called. This might not always be the case – for complex transformations and/or complex integrations – but in this case, it seems fine.

---

In practice, I'm not sure what's right: I still haven't worked on a team where there were enough tests to live up to my theory: of stopping me from changing functionality I did not intend to. I also probably spend too much time writing tests.

In this case, I'm fairly confident I'd rather have the more encompassing test, however.